ruma-api-macros/src/api/request.rs

```diff
@@ -592,6 +592,214 @@ impl Request {
+                        #percent_encoding::utf8_percent_encode(
+                            &self.#path_var.to_string(),
+                            #percent_encoding::NON_ALPHANUMERIC,
+                        )
+                    });
+                    format_string.replace_range(start_of_segment..end_of_segment, "{}");
+                }
+
+                quote! {
+                    format_args!(#format_string, #(#format_args),*)
+                }
+            };
+
+            let path_fields =
+                path_string[1..].split('/').enumerate().filter(|(_, s)| s.starts_with(':')).map(
+                    |(i, segment)| {
+                        let path_var = &segment[1..];
+                        let path_var_ident = Ident::new(path_var, Span::call_site());
+                        quote! {
+                            #path_var_ident: {
+                                let segment = path_segments[#i].as_bytes();
+                                let decoded = #ruma_api::try_deserialize!(
+                                    self,
+                                    #percent_encoding::percent_decode(segment)
+                                        .decode_utf8(),
+                                );
+
+                                #ruma_api::try_deserialize!(
+                                    self,
+                                    ::std::convert::TryFrom::try_from(&*decoded),
+                                )
+                            }
+                        }
+                    },
+                );
+
+            (format_call, quote! { #(#path_fields,)* })
+        } else {
+            (quote! { metadata.path.to_owned() }, TokenStream::new())
+        }
+    }
+
+    /// The function determines the type of query string that needs to be built
+    /// and then builds it using `ruma_serde::urlencoded::to_string`.
+    fn build_query_string(&self, ruma_api: &TokenStream) -> TokenStream {
+        let ruma_serde = quote! { #ruma_api::exports::ruma_serde };
+
+        if let Some(field) = self.query_map_field() {
+            let field_name = field.ident.as_ref().expect("expected field to have identifier");
+
+            quote!({
+            // This function exists so that the compiler will throw an
+            // error when the type of the field with the query_map
+            // attribute doesn't implement IntoIterator<Item = (String, String)>
+            //
+            // This is necessary because the ruma_serde::urlencoded::to_string
+            // call will result in a runtime error when the type cannot be
+            // encoded as a list key-value pairs (?key1=value1&key2=value2)
+            //
+            // By asserting that it implements the iterator trait, we can
+            // ensure that it won't fail.
+            fn assert_trait_impl<T>(_: &T)
+            where
+                T: ::std::iter::IntoIterator<Item = (::std::string::String, ::std::string::String)>,
```

Comment:
Looks like this already slightly overflowed the 100 character limit before, but will do so even more after increasing the indentation here. Please split this line like this:

```rust
T: ::std::iter::IntoIterator<
    Item = /* ... */,
>,
```